app/Helpers/database/player-achievement.php

```diff
@@ -94,9 +67,47 @@ function unlockAchievement(User $user, int $achievementId, bool $isHardcore): ar
         postActivity($user, ActivityType::UnlockedAchievement, $achievement->ID, (int) $isHardcore);
     }
 
-    $completion = getUnlockCounts($achievement->GameID, $user->username, $isHardcore);
-    if (array_key_exists('NumAwarded', $completion)) {
-        $retVal['AchievementsRemaining'] = $completion['NumAch'] - $completion['NumAwarded'];
+    $playerGame = PlayerGame::where('user_id', $user->id)
+        ->where('game_id', $achievement->GameID)
+        ->first();
+
+    // Quick adjustments so we can return correct values in the response. They'll be fully
+    // corrected by UpdatePlayerGameMetrics triggered by an asynchronous UnlockPlayerAchievementJob.
+    if ($isHardcore && !$hasHardcore) {
+        $user->RAPoints += $achievement->Points;
+        if ($hasRegular) {
+            $user->RASoftcorePoints -= $achievement->Points;
+        }
+        $user->save();
+
+        if ($playerGame) {
+            $playerGame->points_hardcore += $achievement->Points;
+            $playerGame->achievements_unlocked_hardcore++;
+
+            if ($hasRegular) {
+                $playerGame->points -= $achievement->Points;
+                $playerGame->achievements_unlocked--;
+            }
+
+            $playerGame->save();
+        }
+    } elseif (!$isHardcore && !$hasRegular) {
+        $user->RASoftcorePoints += $achievement->Points;
+        $user->save();
+
+        if ($playerGame) {
+            $playerGame->points += $achievement->Points;
+            $playerGame->achievements_unlocked++;
+            $playerGame->save();
+        }
+    }
+
+    if ($playerGame) {
+        if ($isHardcore) {
+            $retVal['AchievementsRemaining'] = $achievement->game->achievements_published - $playerGame->achievements_unlocked_hardcore;
+        } else {
+            $retVal['AchievementsRemaining'] = $achievement->game->achievements_published - $playerGame->achievements_unlocked;
+        }
```

Comment:
I'm not sure that we should do any synchronous updates here. the unlock job does all of that, the unlockAchievement method was bound to be removed.
That stats are out of sync for a while it ok, the optimistic response is enough for clients.

This would allow for thousands of unlock requests to not overwhelm the database again as it did in the past.

Reply:
the idea is to getting rid of most of these global helper functions in the long run.
using this unlock method to write anything to the db defeats the purpose of moving most of the write operations to queued jobs.

Reply:
The client is expecting an immediate response of the player's new score and how many achievements they have left to unlock. The optimistic response is fine if only one achievement is unlocked, but there are many times where multiple achievements get unlocked at the same time (defeat final boss, defeat final boss without continuing, have all items after defeating final boss, etc). In that case, to show the mastery notification on the client, the "AchievementsRemaining" must be 0 on one of the calls. If they all optimistically subtract 1 from the `player_games->achievements_unlocked` counter, the client will see "2 remaining" for all three requests and not show the notification.

Similarly, the client could receive "base points+achievement1 points", "base points+achievement2 points" and "base points+achievement3 points" and be out-of-sync with the server. I'm less concerned about this as few people actually look at their points in the client, but we need to do something to ensure the mastery is shown correctly. 

Maybe we could do something ephemeral with redis where it caches the number of pending unlocks?

Reply:
hrm, i was not aware that mastery popups are in play here, too. can we reduce the writing to player_games?
that'd leave points out of it and use player_games as the only reference to keep track. i don't think redis would make this much easier to maintain.